master/buildbot/worker/latent.py

```diff
@@ -101,27 +104,47 @@ def substantiate(self, sb, build):
 
     def _substantiate(self, build):
         # register event trigger
-        d = self.start_instance(build)
+        try:
+            d = self.start_instance(build)
+        except Exception:
+            # if start_instance crashes without defer, we still handle the cleanup
+            d = defer.fail(failure.Failure())
 
         def start_instance_result(result):
             # If we don't report success, then preparation failed.
+            # we let the errback handle the issue
             if not result:
+                # this behaviour is kept as compatibility, but it is better
+                # to just errback with a workable reason
                 msg = "Worker does not want to substantiate at this time"
-                self._substantiation_notifier.notify(LatentWorkerFailedToSubstantiate(self.name, msg))
-                return None
+                return LatentWorkerFailedToSubstantiate(self.name, msg)
             return result
 
         def clean_up(failure):
+            self.startSubstantiationRetryTimer()
             if self.missing_timer is not None:
                 self.missing_timer.cancel()
             self._substantiation_failed(failure)
             # swallow the failure as it is given to notified
             return None
-        d.addCallbacks(start_instance_result, clean_up)
+        d.addCallback(start_instance_result)
+        d.addErrback(clean_up)
         return d
 
+    def startSubstantiationRetryTimer(self):
+        self.retry_timer = self.master.reactor.callLater(self.retry_timeout, self.retrySubstantiation)
+        # next time we will wait twice
+        # timer reset on substanciation success
+        if self.retry_timeout < self.retry_max_timeout:
```

Comment:
This would allow the timeout to be greater than the max timeout. If it's ever in the range of (retry_max_timeout/2, retry_max_timeout), when doubled it will then be greater than retry_max_timeout. I think it should be doubled and then clamped to the maximum.